service/src/main/java/bio/terra/workspace/service/job/WorkspaceActivityLogHooks.java

```diff
@@ -0,0 +1,124 @@
+import bio.terra.workspace.db.WorkspaceActivityLogDao;
+import bio.terra.workspace.db.WorkspaceDao;
+import bio.terra.workspace.db.exception.WorkspaceNotFoundException;
+import bio.terra.workspace.db.model.DbWorkspaceActivityLog;
+import bio.terra.workspace.service.resource.controlled.flight.delete.DeleteControlledResourceFlight;
+import bio.terra.workspace.service.resource.controlled.model.ControlledResource;
+import bio.terra.workspace.service.resource.exception.ResourceNotFoundException;
+import bio.terra.workspace.service.workspace.flight.DeleteAzureContextFlight;
+import bio.terra.workspace.service.workspace.flight.DeleteGcpContextFlight;
+import bio.terra.workspace.service.workspace.flight.WorkspaceFlightMapKeys;
+import bio.terra.workspace.service.workspace.flight.WorkspaceFlightMapKeys.ResourceKeys;
+import bio.terra.workspace.service.workspace.model.CloudPlatform;
+import bio.terra.workspace.service.workspace.model.OperationType;
+import java.util.Optional;
+import java.util.UUID;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class WorkspaceActivityLogHooks implements StairwayHook {
+  private static final Logger logger = LoggerFactory.getLogger(StairwayHook.class);
+
+  private final WorkspaceActivityLogDao activityLogDao;
+  private final WorkspaceDao workspaceDao;
+  private final ResourceDao resourceDao;
+
+  private static final String DELETE_GCP_CONTEXT_FLIGHT = DeleteGcpContextFlight.class.getName();
+  private static final String DELETE_AZURE_CONTEXT_FLIGHT =
+      DeleteAzureContextFlight.class.getName();
+
+  private static final String DELETE_CONTROLLED_RESOURCE_FLIGHT =
+      DeleteControlledResourceFlight.class.getName();
+
+  public WorkspaceActivityLogHooks(
+      WorkspaceActivityLogDao activityLogDao, WorkspaceDao workspaceDao, ResourceDao resourceDao) {
+    this.activityLogDao = activityLogDao;
+    this.workspaceDao = workspaceDao;
+    this.resourceDao = resourceDao;
+  }
+
+  @Override
+  public HookAction endFlight(FlightContext context) {
+    logger.info(
+        String.format("endFlight %s: %s", context.getFlightClassName(), context.getFlightStatus()));
+    var workspaceId =
+        context.getInputParameters().get(WorkspaceFlightMapKeys.WORKSPACE_ID, String.class);
+    var operationType =
+        context
+            .getInputParameters()
+            .get(WorkspaceFlightMapKeys.OPERATION_TYPE, OperationType.class);
+    UUID workspaceUuid = UUID.fromString(workspaceId);
+    if (context.getFlightStatus() == FlightStatus.SUCCESS) {
+      activityLogDao.writeActivity(
+          workspaceUuid, new DbWorkspaceActivityLog().operationType(operationType));
+      return HookAction.CONTINUE;
+    }
+    // For DELETE flight, even if the flight failed, database deletion cannot be undone, so it's
+    // still deleted from the database and we should log that as a changed activity.
+    if (operationType == OperationType.DELETE) {
+      if (isWorkspaceDeleted(workspaceUuid)) {
```

Comment:
Let's make this future-proof. If someone adds a new Delete flight, they're not going to remember to update this.

Here's a suggestion that adds logging, and also makes this more modular.

```
if (operationType == OperationType.DELETE) {
  if flight is WorkspaceDeleteFlight
    maybeLogWorkspaceDeletion()
  else if flight is DeleteGcpContext or DeleteAzureContext
    maybeLogCloudContextDeletion
  else if flight is DeleteControlledResourceFlight
    maybeLogControlledResourceFlightDeletion
  else
    throw "Unknown deletion flight: %s"


maybeLogWorkspaceDeletion() {
  if (isWorkspaceDeleted()) {
    logger.info("Writing workspace {} deletion to activity log");
  } else {
    logger.info("Workspace {} not deleted; not writing deletion to activity log");
  }
}
```